### PR TITLE
date_format should not suggest enabling incompatibleDateFormats for formats we cannot support

### LINKED ISSUE #2517: [FEA] date_format should not suggest enabling incompatibleDateFormats for formats we cannot support
**Is your feature request related to a problem? Please describe.**
When using a date_format expression with a format string that is not compatible with GPU, such as 'F' we see this EXPLAIN output:

```
cannot run on GPU because incompatible format 'F'. Set spark.rapids.sql.incompatibleDateFormats.enabled=true to force onto GPU
```

However, setting this flag will not make this run on GPU because it is unsupported.

**Describe the solution you'd like**
If we know we can't support it on GPU, we should not suggest trying to make it run on GPU.

**Describe alternatives you've considered**
None

**Additional context**
None

### PR DESCRIPTION
**Pull request by andygrove**
Signed-off-by: Andy Grove <andygrove@nvidia.com>

Closes #2517.

### CODE REVIEW

File: integration_tests/src/main/python/date_time_test.py [hunk 1]
```diff
@@ -214,3 +214,36 @@ def test_string_unix_timestamp(data_gen, date_form):
...
+
+unsupported_date_formats = ['F']
+@pytest.mark.parametrize('date_format', unsupported_date_formats, ids=idfn)
+@pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
+@pytest.mark.xfail(reason='cuDF does not support these formats')
+def test_date_format(data_gen, date_format):
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : unary_op_df(spark, data_gen).selectExpr("date_format(a, '{}')".format(date_format)))
+
+unsupported_date_formats_force = ['F']
```
Comment (wjxiz1992) [inline integration_tests/src/main/python/date_time_test.py:226]: This seems to be duplicated?

File: integration_tests/src/main/python/date_time_test.py [hunk 2]
```diff
@@ -214,3 +214,36 @@ def test_string_unix_timestamp(data_gen, date_form):
...
+
+maybe_supported_date_formats = ['dd-MM-yyyy']
+@pytest.mark.parametrize('date_format', maybe_supported_date_formats, ids=idfn)
+@pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
+@pytest.mark.xfail(reason='These formats can be parsed but are not certified as supported')
+def test_date_format(data_gen, date_format):
+    assert_gpu_and_cpu_are_equal_collect(
+        lambda spark : unary_op_df(spark, data_gen).selectExpr("date_format(a, '{}')".format(date_format)))
+
+maybe_supported_date_formats_force = ['dd-MM-yyyy']
```
Comment (wjxiz1992) [inline integration_tests/src/main/python/date_time_test.py:243]: Same question here

File: integration_tests/src/main/python/date_time_test.py (hunk 1 above)
Comment (andygrove) [reply integration_tests/src/main/python/date_time_test.py:226]: The second version of the test specifies `conf = {"spark.rapids.sql.incompatibleDateFormats.enabled": "true"}`

Comment (wjxiz1992) [reply integration_tests/src/main/python/date_time_test.py:226]: Oh, sorry, I mean the variable `unsupported_date_formats_force` is assigned but not used.

File: integration_tests/src/main/python/date_time_test.py [hunk 3]
```diff
@@ -214,3 +214,36 @@ def test_string_unix_timestamp(data_gen, date_form):
 def test_date_format(data_gen, date_format):
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark : unary_op_df(spark, data_gen).selectExpr("date_format(a, '{}')".format(date_format)))
+
+unsupported_date_formats = ['F']
+@pytest.mark.parametrize('date_format', unsupported_date_formats, ids=idfn)
+@pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
+@pytest.mark.xfail(reason='cuDF does not support these formats')
```
Comment (revans2) [inline integration_tests/src/main/python/date_time_test.py:221]: Why xfail and not test that we fall back to the CPU?

File: integration_tests/src/main/python/date_time_test.py (hunk 1 above)
Comment (andygrove) [reply integration_tests/src/main/python/date_time_test.py]: Thanks, this is fixed now. There were numerous other issues with the tests that are also resolved now.

File: integration_tests/src/main/python/date_time_test.py (hunk 3 above)
Comment (andygrove) [reply integration_tests/src/main/python/date_time_test.py]: Thanks, this is fixed now.

**Comment by andygrove** [comment]
build

**Comment by sameerz** [comment]
@andygrove can we retarget this to 21.08?

**Review by revans2 — APPROVED** [review]

**Comment by sameerz** [comment]
Moving to branch-21.08 after @Salonijain27 confirmed with the PR author.

File: integration_tests/src/main/python/date_time_test.py [hunk 4]
```diff
@@ -214,3 +214,36 @@ def test_string_unix_timestamp(data_gen, date_form):
 def test_date_format(data_gen, date_format):
     assert_gpu_and_cpu_are_equal_collect(
         lambda spark : unary_op_df(spark, data_gen).selectExpr("date_format(a, '{}')".format(date_format)))
+
+unsupported_date_formats = ['F']
+@pytest.mark.parametrize('date_format', unsupported_date_formats, ids=idfn)
+@pytest.mark.parametrize('data_gen', date_n_time_gens, ids=idfn)
+@allow_non_gpu('ProjectExec,date_format,Alias,DateFormatClass,Literal,Cast')
```
Comment (revans2) [inline integration_tests/src/main/python/date_time_test.py:221]: nit: I don't think that `date_format` needs to be in the `allow_non_gpu` tag.

**Review by revans2 — APPROVED** [review]

File: integration_tests/src/main/python/date_time_test.py (hunk 4 above)
Comment (andygrove) [reply integration_tests/src/main/python/date_time_test.py]: This has been updated now

**Review by revans2 — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Pull request merged**